File: src/main/java/com/baulsupp/oksocial/services/google/DiscoveryIndex.java
package com.baulsupp.oksocial.services.google;

import com.baulsupp.oksocial.util.JsonUtil;
import com.google.common.collect.Lists;
import com.google.common.io.Resources;
import java.io.IOException;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import static java.util.Collections.emptyList;
import static java.util.Comparator.comparing;
import static java.util.stream.Collectors.toList;

/**
 * API URL -> Discovery URL
 */
public class DiscoveryIndex {
  private final Map<String, List<String>> map;

  public DiscoveryIndex(Map<String, List<String>> map) {
    this.map = map;
  }

  public static DiscoveryIndex loadStatic() throws IOException {
    URL url = DiscoveryIndex.class.getResource("index.json");

    String definition = Resources.toString(url, StandardCharsets.UTF_8);

    return DiscoveryIndex.parse(definition);
  }

  public static DiscoveryIndex parse(String definition) throws IOException {
    Map m = JsonUtil.map(definition);
    return new DiscoveryIndex(m);
  }

  /**
   * Exact search
   */
  public List<String> getDiscoveryUrlForApi(String api) {
    return Optional.ofNullable(map.get(api)).orElse(Lists.newArrayList());
  }

  /**
   * Prefix search (returns longest)
   */
  public List<String> getDiscoveryUrlForPrefix(String prefix) {
    return map.entrySet()
        .stream()
        .filter(s1 -> match(prefix, s1.getKey()))
        .flatMap(s -> s.getValue().stream())
        .collect(toList());
  }

  boolean match(String prefix, String indexKey) {
    return indexKey.startsWith(prefix) || prefix.startsWith(indexKey);
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/services/google/DiscoveryIndex.kt
package com.baulsupp.oksocial.services.google

import com.baulsupp.oksocial.output.util.JsonUtil

/*
 * API URL -> Discovery URL
 */
class
DiscoveryIndex(private val map: Map<String, List<String>>) {

  /*
   * Exact search
   */
  fun getDiscoveryUrlForApi(api: String): List<String> = map[api] ?: listOf()

  /*
   * Prefix search (returns longest)
   */
  fun getDiscoveryUrlForPrefix(prefix: String): List<String> = map.entries
          .filter { s1 -> match(prefix, s1.key) }
          .flatMap { s -> s.value }

  internal fun match(prefix: String, indexKey: String): Boolean =
          indexKey.startsWith(prefix) || prefix.startsWith(indexKey)

  companion object {
    fun loadStatic(): DiscoveryIndex =
      DiscoveryIndex::class.java.getResource("index.json")!!.let { DiscoveryIndex.parse(it.readText()) }

    fun parse(definition: String): DiscoveryIndex {
      val m = JsonUtil.map(definition)
      return DiscoveryIndex(m as Map<String, List<String>>)
    }
  }
}
